Tests/SwiftDiagnosticsTest/DiagnosticTestingUtils.swift

```diff
@@ -0,0 +1,221 @@
+///
+/// ### Example
+///
+/// In the following test code snippet, the emojis 1️⃣ and 2️⃣ are used as location markers:
+///
+/// ```swift
+/// func foo() -> Int {
+///   if 1️⃣1 != 0 2️⃣{
+///     return 0
+///   }
+///   return 1
+/// }
+/// ```
+typealias LocationMarker = String
+
+/// Represents a descriptor for constructing a diagnostic in testing.
+struct DiagnosticDescriptor {
+  /// The marker pointing to location in source code.
+  let locationMarker: LocationMarker
+
+  /// The ID associated with the message, used for categorizing or referencing it.
+  let id: MessageID
+
+  /// The textual content of the message to be displayed.
+  let message: String
+
+  /// The severity level of the diagnostic message.
+  let severity: DiagnosticSeverity
+
+  /// The syntax elements to be highlighted for this diagnostic message.
+  let highlight: [Syntax]  // TODO: How to create an abstract model for this?
+
+  /// Descriptors for any accompanying notes for this diagnostic message.
+  let noteDescriptors: [NoteDescriptor]
+
+  /// Descriptors for any Fix-Its that can be applied for this diagnostic message.
+  let fixIts: [FixIt]  // TODO: How to create an abstract model for this?
+
+  /// Initializes a new `DiagnosticDescriptor`.
+  ///
+  /// - Parameters:
+  ///   - locationMarker: The marker pointing to location in source code.
+  ///   - id: The message ID of the diagnostic.
+  ///   - message: The textual message to display for the diagnostic.
+  ///   - severity: The severity level of the diagnostic. Default is `.error`.
+  ///   - highlight: The syntax elements to be highlighted. Default is an empty array.
+  ///   - noteDescriptors: An array of note descriptors for additional context. Default is an empty array.
+  ///   - fixIts: An array of Fix-It descriptors for quick fixes. Default is an empty array.
+  init(
+    locationMarker: LocationMarker,
+    id: MessageID = MessageID(domain: "test", id: "conjured"),
+    message: String,
+    severity: DiagnosticSeverity = .error,
+    highlight: [Syntax] = [],
+    noteDescriptors: [NoteDescriptor] = [],
+    fixIts: [FixIt] = []
+  ) {
+    self.locationMarker = locationMarker
+    self.id = id
+    self.message = message
+    self.severity = severity
+    self.highlight = highlight
+    self.noteDescriptors = noteDescriptors
+    self.fixIts = fixIts
+  }
+
+  /// Creates a ``Diagnostic`` instance from a given ``DiagnosticDescriptor``, syntax tree, and location markers.
+  ///
+  /// - Parameters:
+  ///   - tree: The syntax tree where the diagnostic is rooted.
+  ///   - markers: A dictionary mapping location markers to their respective offsets in the source code.
+  ///   - file: The file where the test is located, used for reporting failures.
+  ///   - line: The line where the test is located, used for reporting failures.
+  ///
+  /// - Returns: A ``Diagnostic`` instance populated with details from the ``DiagnosticDescriptor``, or `nil` if it fails.
```

Comment:
Instead of silently failing and not generating a `Diagnostic` for a `DiagnosticDescriptor` if it fails, what do you think about throwing an error instead? I think that should make it a lot clearer what’s going wrong if you write eg. a `DiagnosticDescriptor` with a marker that doesn’t exist in the source code.

Reply:
I believe that implementing the `assert` function in this way would indeed make the code clearer, but I'm not entirely convinced that the end result would be more transparent to the users of this high-level API. 

My primary motivation behind this implementation approach was to avoid interrupting the assertion process unnecessarily. The `assertAnnotated` function takes a collection of `diagnosticDescriptors` as an argument. I chose to return `nil` while also invoking `XCTFail` so that we could iterate over all elements, inform the user only about the missing ones, and apply the existing ones to provide as much assertion validation as possible. For instance, if only one location marker out of four is missing, the user will end up with the following result:

![Screenshot 2023-10-11 at 7 18 50 PM](https://github.com/apple/swift-syntax/assets/44930823/d1db24f0-6def-4155-9dd8-1a054c67f4e7)

I believe this approach benefits end clients. If I'm not mistaken, it wouldn't work in the same way if we chose to throw an error, would it?

Reply:
Oh, I missed the fact that we always emitted an `XCTFail` if we return `nil`.

And I think me missing that is a reason why I would prefer to `throw`. My thinking is that if you are missing one of the diagnostic markers in your test input, there’s something wrong with the test case itself (not just a runtime failure) and thus early exiting is quite alright.

As it stands right now, every time I would read this code, I would wonder the same thing that I originally commented: 
- Are we silently dropping one of our test case assertions here? That seems dangerous
- Looks at implementation/documentation: Oh, no, it’s fine we emit an `XCTFail` in that case.
- Now, what was I doing originally? Needs to restore mental context after side-stepping.

Reply:
You make some solid points. You got me—I'm convinced.

I'll swap out the `nil` and `XCTFail` combo for a good ol' error throw. Thanks for the nudge!